### ISSUE TITLE
Debian releases with a predicatble version

### ISSUE THREAD

**Issue by evandam** [buildkite/agent#1348]
Hi folks,

Whenever I upgrade versions of the Buildkite agent on Ubuntu, I find myself spinning up a VM to attempt to change the version, watch it fail, and then manually log in to find the full version of the Debian package with an `apt list buildkite-agent`.

For example, the last number is always incrementing and inconsistent, possibly a build number?

```
root@default-ubuntu-1804:~# apt list buildkite-agent -a
Listing... Done
buildkite-agent/unknown 3.26.0-3674 amd64
buildkite-agent/unknown 3.25.0-3651 amd64
buildkite-agent/unknown 3.24.0-3587 amd64
buildkite-agent/unknown 3.23.1-3540 amd64
buildkite-agent/unknown 3.23.0-3534 amd64
buildkite-agent/unknown 3.22.1-3416 amd64
buildkite-agent/unknown 3.22.0-3403 amd64
buildkite-agent/unknown 3.21.1-3397 amd64
buildkite-agent/unknown 3.21.0-3391 amd64
...
```

Is it possible to make this something consistent like `3.26.0-1`?

**Comment by scalp42** [comment]
Can confirm, it's not deterministic at all and a challenge for us.

**Comment by sj26** [comment]
Sorry folks, this is a build number, and used to disambiguate packages built at the same version before they are promoted to stable.

Why do you need to know this value? Is there another way we can help you solve your problem?

**Comment by yob** [comment]
It's a fair point - official debian packaging policy is for the `-1` to be the version of the packaging within an upstream release and resetting it to `-1` is the norm for packages in the debian/ubuntu repositories.

The way it's used here is fairly baked into our release process though and changing that may not be trivial. It'd be good to understand more about why it causes problems - maybe you need the exact package URL to use in scripts or config?

**Comment by evandam** [comment]
It's mostly around ease of upgrading versions. Outside of your release process, the number really doesn't mean anything externally to customers.

We pin versions of the buildkite-agent package, so I need to get the exact version of the package to upgrade it. With other packages, I can typically rely on using `x.y.z-1` or some standard pattern. With Buildkite, however, there is no way to get this easily besides logging into an instance/VM and searching for the new package to drop into our configuration management.

**Comment by sj26** [comment]
Hmm, if we didn't use build number then it would likely be `-15` or `-23` or `-5` because we rarely release the first built package.

So the problem is discoverability? What if we had an index of the apt packages with full versions?

**Comment by evandam** [comment]
Sure, an index of the repo would work too 👌

**Comment by yob** [comment]
> Hmm, if we didn't use build number then it would likely be -15 or -23 or -5 because we rarely release the first built package.

This is true in the official debian repositories as well. foo `1.2.3-1`, `1.2.3-2` and `1.2.3-3` might all be uploaded to the unstable distribution, and only the final one (`1.2.3-3`) might be promoted to the stable distribution. For anyone tracking stable, `1.2.3-3` is the first package for `1.2.3` that they'll see.

> So the problem is discoverability? What if we had an index of the apt packages with full versions?

Could we expand the [agent releases URL](https://buildkite.com/agent/releases/latest?platform=linux) to include debian package version/info in some cases?

**Comment by sj26** [comment]
> > So the problem is discoverability? What if we had an index of the apt packages with full versions?
>
> Could we expand the agent releases URL to include debian package version/info in some cases?

Hmm, I think the information comes from different places. The agent releases url is built from GitHub releases which do not have apt packages.

I was thinking more an index page on https://apt.buildkite.com, generated from the apt repository directly. I'm sure there are tools to do this.
